don't merge PCN (resolves #10193)

baiti opened the pull request:
see referenced issue #10193.

In a nutshell: Provide a preference setting to tell cgeo not to merge personal cache notes data coming down from the provider with what cgeo already might have stored locally. The datamodel behind this is that there is only one authoritative source for that data and that is in the provider's database. On refresh, take what comes down verbatim. If local changes are made, make sure to push them to the provider manually. The functionality to push, is already there, even for lists.

fm-sys reviewed (commented):
Looks good codewise. However we should probably add a hint to the description string that it does only work for GC premium members and disable the setting if it's not the case. 

Furthermore, we have briefly discussed in our yesterdays meeting that we should avoid introducing new settings for the next months, to make migration to androidX easier (see  #11298). But at the end, @SchoolGuy has to decide if it's ok to introduce new settings for the time being.

baiti commented:
> #11298

> Looks good codewise. However we should probably add a hint to the description string that it does only work for GC premium members and disable the setting if it's not the case.

I was wondering, isn't it just a no-op for basic members? In other words, assume this PR gets merged the way it is now, will a basic user be able to observe any difference if he/she sets or resets that flag?

> 
> Furthermore, we have briefly discussed in our yesterdays meeting that we should avoid introducing new settings for the next months, to make migration to androidX easier (see #11298). But at the end, @SchoolGuy has to decide if it's ok to introduce new settings for the time being.

Ahh ok. Didn't know that this work is in progress. Of course, this PR here can wait until the migration has been done. Reworking it to the androidX model should be rather simple then.

bekuno commented on main/res/values/strings.xml:
```diff
@@ -791,6 +791,8 @@
     <string name="init_summary_bigSmileysOnMap">If enabled big emoticons will be shown for logged caches on the map instead of the cache type icon (requires restart).</string>
     <string name="init_global_wp_extraction_disable">Disable waypoint extraction</string>
     <string name="init_summary_global_wp_extraction_disable">Globally disable waypoint extraction from personal cache notes. It can still be enabled for individual caches by deactivating the checkbox below the personal cache note</string>
+    <string name="init_prevent_cache_note_merge">Prevent Merging of Personal Cache Notes</string>
+    <string name="init_summary_prevent_cache_note_merge">Prevent merging of personal cache note data from provider with locally stored personal caches notes data</string>
```
Do you really mean "Do not merge" or do you mean "Replace local cache note by the server side cache note"?

baiti replied:
Interesting question. Hard to describe it with a sentence as brief as both variants are.

What is meant is the following: Upon initially loading a geocache object from the provider, take the personal cache notes from the provider unchanged. Further, upon refreshing a geocache object from the provider, take the personal cache notes from the provider unchanged. If a user has made local modifications to the personal notes fielld in cgeo he must push those modifications to the provider otherwise those changes are lost with the next refresh of the object.

Does `Replace local cache note by the server side cache note` transport these semantics sufficiently?

bekuno commented on main/res/values/strings.xml:
```diff
@@ -791,6 +791,8 @@
     <string name="init_summary_bigSmileysOnMap">If enabled big emoticons will be shown for logged caches on the map instead of the cache type icon (requires restart).</string>
     <string name="init_global_wp_extraction_disable">Disable waypoint extraction</string>
     <string name="init_summary_global_wp_extraction_disable">Globally disable waypoint extraction from personal cache notes. It can still be enabled for individual caches by deactivating the checkbox below the personal cache note</string>
+    <string name="init_prevent_cache_note_merge">Prevent Merging of Personal Cache Notes</string>
```
So much uppercase words -> "Prevent merging of personal cache notes" fits better with the other texts.

baiti replied:
I have absolutely no problem to rephrase those words to anything that makes it more readable and easier to understand though.

Reason why so much was uppercase is that in English speaking countries, words in titles are uppercase while they are lowercase in regular text.

baiti replied:
will commit your suggestion: Prevent merging of personal cache notes" fits better with the other texts

baiti replied:
Actually this topic made me thinking. Look at the preference screen:

![Screenshot from 2021-09-14 07-35-37](https://user-images.githubusercontent.com/5296333/133201097-dab46045-bdca-49bb-af7f-e82b7414081d.png)

First of all, you can see that I have made my new entry consistent with the one directly above (as suggested by @bekuno  in the review) with respect to capitalization of the first letter of the words, however, the rules in English writing are different: https://www.webucator.com/article/how-to-capitalize-headings-and-titles/

Consistent with that style we have: 

- Show Friends/Own <-- all first letters uppercase
- Last Details Page <- all first letters uppercase
- Show Direction
- Use Chrome Webview

but:

- Disable waypoint extraction <-- which is not consistent with what the referenced document says
- Prevent merging of personal cache notes <-- does not comply with the rules for headings and titles

I would say, we have to establish some rules and then make it all consistent throughout all views we render in cgeo. If those rule are in sync  with the referenced link above, then those two lines after "but" should be changed to:

 - Disable Waypoint Extraction
 - Prevent Merging of Personal Cache Notes

Lineflyer commented:
My few cents: PN upload is also supported on most OC-nodes:
1. The behavior should work for all connectors
2. The texts for the setting could IMHO be more direct, to make perfect clear, that if this is activated the server note will overwrite whats stored locally without merging or combining.

baiti commented:
> My few cents: PN upload is also supported on most OC-nodes:
> 
> 1. The behavior should work for all connectors

Yes

> 2. The texts for the setting could IMHO be more direct, to make perfect clear, that if this is activated the server note will overwrite whats stored locally without merging or combining.

Feel free to make a suggestion for the wording

SchoolGuy commented:
So my two cents on this PR are the following: If I haven't gotten the my PR working by the end of this month I think it doesn't make a difference if I rebase again on `master`. So please be patient until then.

baiti commented:
Changed wording as requested by @bekuno and as suggested by @Lineflyer.

Is there anything else that blocks this PR at this time?

bekuno commented on main/res/values/strings.xml:
```diff
@@ -791,6 +791,8 @@
     <string name="init_summary_bigSmileysOnMap">If enabled big emoticons will be shown for logged caches on the map instead of the cache type icon (requires restart).</string>
     <string name="init_global_wp_extraction_disable">Disable waypoint extraction</string>
     <string name="init_summary_global_wp_extraction_disable">Globally disable waypoint extraction from personal cache notes. It can still be enabled for individual caches by deactivating the checkbox below the personal cache note</string>
+    <string name="init_prevent_cache_note_merge">Prevent merging of Personal Cache Notes</string>
+    <string name="init_summary_prevent_cache_note_merge">If this setting is enabled c:geo will delete the locally stored note permanently and replace it by the personal note on the server. No merging will be done</string>
```
We should use "note" instead of "notes" (singular form) in all texts, imho.
E.g., iIt is used in this way at the geocaching.com website.

The "." at the end of the last sentences are missing (for init_summary_global_wp_extraction_disable and init_summary_prevent_cache_note_merge).

baiti replied:
ok, done. Both strings you mentioned now have terminating dots.

baiti commented:
Is there anything I can or must do in order to get this PR merged? It is now open since 9/13 and @fm-sys assigned a "Work in progress / do not merge". Is my assumption correct that it is also @fm-sys who ultimately decides to remove that flag? Who else would be removing it instead? It seems I still don't quite understand the processes in open source development.

If it just so happens that this thing sits in the queue because nobody except me actually has bought into the usefulness of this function, then pls speak up and let me know and I can discard the branch and potentially focus on other, more interesting work.

eddiemuc commented:
Scrolling quickly through this PRs history I think the sole reason why this is still marked as "do not merge" is the dependency on @SchoolGuy s work.

@SchoolGuy ok to merge here?

Lineflyer reviewed (approved):
AFAIR we waited for @SchoolGuy regarding refactoring of our settings. So lets wait for his feedback. If settings migration needs significantly more time, we should IMHO merge this to get this done.

baiti commented:
> AFAIR we waited for @SchoolGuy regarding refactoring of our settings. So lets wait for his feedback. If settings migration needs significantly more time, we should IMHO merge this to get this done.

Ohh I see. I thought this was completed long ago. 9/14 @SchoolGuy mentioned "end of the month" and I believe he meant end of September. So I thought the PR was blocked because we hadn't resolved some issues e.g. wording of text.

It still says "changes requested" and I don't know how to formally address this. I believe technically it has all been fixed now so it is perhaps just some missing state handling in github?

Lineflyer commented:
retest this please

SchoolGuy commented:
@baiti you are right. I am deeply sorry that this takes ages. I am involved in too many things... However I will finish the PR next week. So please bear with me for one more week.

baiti commented:
I understand, the changes that this PR was blocked on have now been merged into master. I would kindly ask one of you with merge authority to merge this PR now. Thanks much in advance.

SchoolGuy reviewed (approved):
In my eyes this PR is LGTM. Thanks for the patience @baiti !

Lineflyer commented:
@baiti You need to rebase it and resolve conflicts for the `preferences.xml` as it seems.

baiti commented:
> @baiti You need to rebase it and resolve conflicts for the `preferences.xml` as it seems.

Done! ... but the results don't really look convincing to me. 538 commits? 1090 files changed?

baiti commented:
... besides the mess, the changes I **wanted** to make are nowhere to be seen. I guess I'd have to redo everything from scratch. So essentially, I give up. It is just not worth it. The function in question is perhaps anyhow only used by a handful of users, mostly those I am regular on tour with, folks I may just have talked into the use case.

fm-sys commented:
It seems, commit https://github.com/cgeo/cgeo/commit/231e3bdf9ff104713cbd0fbdd7db206f39d0cad5 was the last state, where everything was fine. I can easily restore that state and try the rebase the PR again. But you have to decide whether you still want that setting at all...

baiti commented:
> It seems, commit [231e3bd](https://github.com/cgeo/cgeo/commit/231e3bdf9ff104713cbd0fbdd7db206f39d0cad5) was the last state, where everything was fine. We can easily restore that state and try the rebase the PR again. But you have to decide whether you still want that setting at all...

In the interest of fixing it I did `git revert 52bec4c7ab98b4b7820b89b77efa9fdf0cdeddc4 -m 2` followed by a `git push`. But it seems nothing has changed.

I made those changes 3 months ago and to be honest, I don't remember any of it at this time so if I had to recreate the PR, I would need to startover from scratch and even though this is a rather simple change, actually a single `if () {}` statement based on a preference setting, it would take me hours to recreate it as I am not working with the code base on a regular basis.

fm-sys commented:
You can try a `git reset --hard 231e3bdf9ff104713cbd0fbdd7db206f39d0cad5` followed by a force push to get the previous state.

Lineflyer commented:
No need to redo your work as long as its preserved in some commits on your machine.
In such cases I would go the way to create a new feature branch based on current `master` and cherry-pick my original feature commit(s) to that new branch (which will then include conflict resolution). Then you should have a clean branch with proper base to issue a new PR.

baiti commented:
> git reset --hard 231e3bdf9ff104713cbd0fbdd7db206f

Done

> No need to redo your work as long as its preserved in some commits on your machine. In such cases I would go the way to create a new feature branch based on current `master` and cherry-pick my original feature commit(s) to that new branch (which will then include conflict resolution). Then you should have a clean branch with proper base to issue a new PR.

Since I am still (after a year now) struggling with this git technology, I try to make only small changes, commit those and try to get it merged as soon as possible. I have never really grasped how to rebase more complex changes. And what you say in the paragraph I am quoting, I do not understand a single word of it.

baiti commented:
> No need to redo your work as long as its preserved in some commits on your machine. In such cases I would go the way to create a new feature branch based on current `master` and cherry-pick my original feature commit(s) to that new branch (which will then include conflict resolution). Then you should have a clean branch with proper base to issue a new PR.

Assume I could somehow accomplish what you said. Would that mean we do a new PR based on the new feature branch and throw this one away then?

fm-sys commented:
> Would that mean we do a new PR based on the new feature branch and throw this one away then?

Actually, no need to do that. Commit history already looks clean again. 🎉👏👍

But now we have merge conflicts again for sure. For rebasing, I always use the help of Android Studio. If you need help, I can also do it for you if you like... 

![grafik](https://user-images.githubusercontent.com/64581222/145839045-6fdc16a8-af64-441b-b147-7d27821e6ba0.png)

Lineflyer commented:
> Assume I could somehow accomplish what you said. Would that mean we do a new PR based on the new feature branch and throw this one away then?

If these are the real relevant commits:
![grafik](https://user-images.githubusercontent.com/949669/145839244-f465215b-6729-4ba9-ab56-e44d13038fe1.png)

I would do it like this (written without having git at hand...but hopefully I made no mistake):
git checkout master
git pull upstream master (assuming upstream is the main cgeo repository)
git checkout -b new_approach_PCN (or whatever name you like)
git cherry-pick commitSHA --no-commit (for all of the above commits)
review if all your changes are in
git commit 
git push --set-upstream origin new_approach_PCN
Create pull request from that on Github

If you struggle I can give it a try this evening...only I will not be able to fix the conflict

Lineflyer commented:
Addon:
It can only give a mess if many people propose different ways to continue. :)
@fm-sys is correct. Now everything looks good already besides the conflict in that single file you have to eliminiate.

baiti commented:
> Addon: It can only give a mess if many people propose different ways to continue. :) @fm-sys is correct. Now everything looks good already besides the conflict in that single file you have to eliminiate.

I understand the point with the mess ;) So far I haven't pursued either way. Neither yours, nor fm-sys'.

It says:

```
baiti@dog:~/cgeo$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
baiti@dog:~/cgeo$ git pull upstream master
From https://github.com/cgeo/cgeo
 * branch                master     -> FETCH_HEAD
Already up to date.
baiti@dog:~/cgeo$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
baiti@dog:~/cgeo$ git switch prevent_cache_note_merge 
Switched to branch 'prevent_cache_note_merge'
Your branch is up to date with 'origin/prevent_cache_note_merge'.
baiti@dog:~/cgeo$ 
```
However, if I do a: `git diff master...prevent_cache_note_merge` I get exactlty my changes I made back then in September. And those I have committed and pushed. Nowhere do I see a conflict that I could try to resolve through whatever means.

@fm-sys ... I'd guess your sceeenshot assumes I'd be authorized to make changes to master? I can only say, fortunately that is not the case as I might easily screw up things in a very bad way then.

fm-sys commented:
> @fm-sys ... I'd guess your sceeenshot assumes I'd be authorized to make changes to master? I can only say, fortunately that is not the case as I might easily screw up things in a very bad way then.

It should work without special permissions as well. Only your branch will be updated, not the master branch. No worries, there's (currently) nothing you can break, even if you try to do so... ;-)

Only thing you need to do is to create a alias `upstream` for the c:geo main repository: 
`git remote add upstream git@github.com:cgeo/cgeo.git`
But most probably you already have done that while setting up your git fork.

baiti commented:
> > @fm-sys ... I'd guess your sceeenshot assumes I'd be authorized to make changes to master? I can only say, fortunately that is not the case as I might easily screw up things in a very bad way then.
> 
> It should work without special permissions as well. Only your branch will be updated, not the master branch. No worries, there's (currently) nothing you can break, even if you try to do so... ;-)
> 
> Only thing you need to do is to create a alias `upstream` for the c:geo main repository: `git remote add upstream git@github.com:cgeo/cgeo.git` But most probably you already have done that while setting up your git fork.

I am currently struggling with my local environment not showing any conflicts. How can this be?

`git diff` ... shows nothing
`git status` ... shows nothing

but

`git diff master` ... gets me all the changes in those 1030 files.

Lineflyer commented:
> I am currently struggling with my local environment not showing any conflicts. How can this be?

The conflict is not on your local environment but on upstream `master`.
Anyway the diff commands will not easily show it.

Pragmatic approach to avoid all that complex git stuff ;) would be:
- Take the content of `main/res/xml/preferences.xml` from Github master in raw view
- Go to the branch you use for this PR on your local PC, open this file and replace the content completely with what you got from master
- Apply your change again into that file
- Commit this change as conflict resolutions and push to update this PR

baiti commented:
> > I am currently struggling with my local environment not showing any conflicts. How can this be?
> 
> The conflict is not on your local environment but on upstream `master`. Anyway the diff commands will not easily show it.
> 
> Pragmatic approach to avoid all that complex git stuff ;) would be:
> 
> * Copy the content of `main/res/xml/preferences.xml` from Github master in raw view
> * Go to the branch you use for this PR on your local PC, open this file and replace (paste) the content completely with what you got from master
> * Apply your change again into that file
> * Commit this change as a conflict resolution and push to update this PR

I looked at master's preferences.xml and at my branch's preferences.xml and they are totally different. I cannot just manually take what's in master, copy that into my branch and manually add my changes. It seems the whole structure has changed from having everything in one big file (which is what I still have in my branch) and splitting it all up into multiple smaller files. It seems I need to find the right "smaller file" and insert my changes into that. @SchoolGuy ... you offered help on the rebase. Now would be a good time to jump in ;)

moving-bits commented:
> Edit: It seems I must not touch preferences.xml. Instead, I need to manually merge my changes to preferences.xml into preferences_cachedetails.xml. Whether that is then sufficient, I don't know

The old monolithic `preferences.xml` and `SettingsActivity.java` have been split up into several smaller frament-based files. So if your change needs to go to `preferences_cachedetails.xml` there will be an accompanying java file named `PreferenceCachedetailsFragment.java` where you can put related java code into.

baiti commented:
> > Edit: It seems I must not touch preferences.xml. Instead, I need to manually merge my changes to preferences.xml into preferences_cachedetails.xml. Whether that is then sufficient, I don't know
> 
> The old monolithic `preferences.xml` and `SettingsActivity.java` have been split up into several smaller frament-based files. So if your change needs to go to `preferences_cachedetails.xml` there will be an accompanying java file named `PreferenceCachedetailsFragment.java` where you can put related java code into.

Thx for jumping in. I have done the first part and not (yet) touched the Java file you mentioned, however, I believe I am already screwed up:

```
baiti@dog:~/cgeo$ git status
On branch prevent_cache_note_merge
Your branch and 'origin/prevent_cache_note_merge' have diverged,
and have 531 and 6 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

nothing to commit, working tree clean
baiti@dog:~/cgeo$ 
```

moving-bits commented:
> Your branch and 'origin/prevent_cache_note_merge' have diverged,
> and have 531 and 6 different commits each, respectively.
>   (use "git pull" to merge the remote branch into yours)

This does only mean, that you have 531 commits locally, which you don't have in your online fork of the c:geo repository (`origin`), and six commits over there that you don't have locally. But since you have rebased your local branch against `upstream` (the original c:geo repo) this is to be expected. For me it's currently close to 4000 commits locally that are not in my fork.

If your local branch is technically ok is easy to see if you push a PR from it, as GitHub will color-code the differences that come from your PR (see above on the "Files changed" tab).

baiti commented:
> > Your branch and 'origin/prevent_cache_note_merge' have diverged,
> > and have 531 and 6 different commits each, respectively.
> > (use "git pull" to merge the remote branch into yours)
> 
> This does only mean, that you have 531 commits locally, which you don't have in your online fork of the c:geo repository (`origin`), and six commits over there that you don't have locally. But since you have rebased your local branch against `upstream` (the original c:geo repo) this is to be expected. For me it's currently close to 4000 commits locally that are not in my fork.
> 
> Whether your local branch is technically ok is easy to see if you push a PR from it, as GitHub will color-code the differences that come from your PR (see above on the "Files changed" tab).

Are you suggesting I now just do a `git push` or `git push -f` ?

baiti commented:
```
baiti@dog:~/cgeo$ git push --force-with-lease
Enumerating objects: 63, done.
Counting objects: 100% (63/63), done.
Delta compression using up to 8 threads
Compressing objects: 100% (48/48), done.
Writing objects: 100% (48/48), 3.97 KiB | 3.97 MiB/s, done.
Total 48 (delta 38), reused 0 (delta 0)
remote: Resolving deltas: 100% (38/38), completed with 13 local objects.
To github.com:baiti/cgeo.git
 + 231e3bdf9...ccb59c355 prevent_cache_note_merge -> prevent_cache_note_merge (forced update)
```

... that MAY have worked ... will check

moving-bits commented:
> Are you suggesting I now just do a `git push` or `git push -f` ?

Yes, should work (but only with "-f". As you have changed the base of your commits, a regular `git push` will fail. Adding a "-f" forces git to ignore this.)

baiti commented:
A quick sniff test on the emulator with the preference unchecked (default) and then checked, worked for me so from that perspective I'd say it is ready for merge

baiti commented:
One question though: Is my change in `main/src/cgeo/geocaching/settings/Settings.java` still at the right place or does that have to be moved to some fragment specific file? Functionally it seems to work where it is now but it might violate the new structure of what goes where.

Lineflyer commented:
Looks nice now.
Shall we squash on merging or would you like the extra lesson to squash your commits? Just asking...not mandatory.

baiti commented:
> Looks nice now. Shall we squash on merging or would you like the extra lesson to squash your commits? Just asking...not mandatory.

lol ... I had the squasing lesson multiple times before but you wouldn't be surprised if I told you I don't remember how it goes? So the answer is "yes".

Perhaps the real lesson is: make frequent coomits and squashes such that the procedure stays fresh

Lineflyer commented:
Fair enough...

Then lets just wait until someone can answer your question (as I can't):
> One question though: Is my change in main/src/cgeo/geocaching/settings/Settings.java still at the right place or does that have to be moved to some fragment specific file? Functionally it seems to work where it is now but it might violate the new structure of what goes where.

moving-bits commented:
> One question though: Is my change in `main/src/cgeo/geocaching/settings/Settings.java` still at the right place or does that have to be moved to some fragment specific file? Functionally it seems to work where it is now but it might violate the new structure of what goes where.

`Settings` is fine, hasn't change.

`Settings` is used for storing and retrieving the values stores in our preferences, whereas `SettingsActivity` and the new `PreferenceXXXFragment` files are used for displaying the preferences, reacting to button taps etc. - `SettingsActivity` and the fragments are the UI, and `Settings` is the storage backend.

fm-sys commented on main/res/values/strings.xml:
```diff
@@ -827,7 +827,9 @@
     <string name="init_global_wp_extraction_disable">Disable waypoint extraction</string>
-    <string name="init_summary_global_wp_extraction_disable">Globally disable waypoint extraction from personal cache notes. It can still be enabled for individual caches by deactivating the checkbox below the personal cache note</string>
+    <string name="init_summary_global_wp_extraction_disable">Globally disable waypoint extraction from personal cache notes. It can still be enabled for individual caches by deactivating the checkbox below the personal cache note.</string>
+    <string name="init_prevent_cache_note_merge">Prevent merging of Personal Cache Note.</string>
```
The heading probably shouldn't contain a `.` at the end. Also it seems quite long. Maybe "Overwrite Personal Note"

baiti replied:
Removing the "." makes sense. The "quite long" I am not sure I understand what you are suggesting. Are you suggesting to replace the whole string by literally just "Overwrite Personal Note"? All three words in uppercase then?

baiti replied:
Check it out ....

Lineflyer commented:
As all checks are green and AFAIU the technical points are clarified, I will merge this.
Strings can be further optimized separately if needed. Lets test it in the nightlies.

Thanks @baiti for your patience and sorry for the additional effort and waiting time due to the settings refactoring.

SchoolGuy commented:
@baiti Since I didn't see the conflict in the WebUI I assumed that the PR worked as you did it initially. Sorry that I didn't had time yesterday evening to jump in right away.

baiti commented:
Thx much for all the help. I still don't know why my initial rebase attempt failed so miserably ( 500 commits, 1000 files changed), however, it is truely remarkable how git can recover from such screw ups, **if** one knows the right command lines (`git reset --hard 231e3bdf9ff104713cbd0fbdd7db206f39d0cad5; git push -f`) and the procedures to continue with. I need to take the time and go into all my old PRs and extract all the many help I got with git into some sort of a concise cheat sheet. Perhaps, in the end, I should setup my own dummy repo and play around with it before I try anything on the real repo.

SchoolGuy commented:
@baiti Don't bother doing that. I never got failed by the following collection: https://ohshitgit.com/

baiti commented:
cool! It clearly tells me, others have been in that brown mess before.

baiti commented:
@SchoolGuy ... I went to I guess it was called "Electronica" fair in Stuttgart back in 1994 to visit SuSE and get my first set of Linux CDs.

SchoolGuy commented:
@baiti That mean those are older then I am in total. 😂 We have some of those in our company museum as well. I hope it was a good experience. My private mail is on GH in case you have questions. ;)
